### PR TITLE
Add Github Actions Windows-based CI runner

### PR DESCRIPTION
**Pull request by rleh**
- [x] Enable Github Actions for this repo @salkinium 
- [x] Windows-based runner with lbuild
- [x] Toolchain arm-none-eabi
- [x] ~Install toolchain for AVR~
  - avr-gcc toolchain does not exist on Windows
- [x] Install some "hosted" GCC-Compiler for Windows? MinGW?
- [x] ~Install Doxypress~
- [x] Run some tests
- [x] Run more tests
- [x] ~Test documentation build~
- [x] Fix errors

### CODE REVIEW

**Comment by salkinium** [comment]
Actions enabled.

**Comment by rleh** [comment]
Looks like I somehow need to trigger Github Actions (e.g. by pushing).

Here is a log: https://github.com/rleh/modm/runs/1074401967

```
shell: C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"
  env:
    pythonLocation: C:\hostedtoolcache\windows\Python\3.8.5\x64
==========================================================================================
Generating: stm32f1_discovery\blink
>> modm: Recomputing device cache...

ERROR: In 'modm:docs':


Traceback (most recent call last):

  File "c:\hostedtoolcache\windows\python\3.8.5\x64\lib\encodings\cp1252.py", line 23, in decode

    return codecs.charmap_decode(input,self.errors,decoding_table)[0]

UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 2136: character maps to <undefined>

[...]
```

**Comment by salkinium** [comment]
Possibly related to this from the install guide:

> If you get a UnicodeEncodeError when calling lbuild, you may need to add export LC_ALL=C.UTF-8 to your .bashrc. Consult the internet until python3 -c "import sys; print(sys.stdout.encoding)" returns UTF-8.

**Comment by rleh** [comment]
>Possibly related to this from the install guide:
>
>> If you get a UnicodeEncodeError when calling lbuild, you may need to add export LC_ALL=C.UTF-8 to your .bashrc. Consult the internet until python3 -c "import sys; print(sys.stdout.encoding)" returns UTF-8.


https://github.community/t/how-to-support-chinese-characters-in-windows-virtual-enviroment/17775/2: `env: PYTHONIOENCODING="utf-8"`

https://github.com/rleh/modm/runs/1075241848
Encoding (`print(sys.stdout.encoding)`) is now `uft-8`, but the lbuild error stays the same.

**Comment by rleh** [comment]
Same issue as #121, but it seems to be fixed in modm-io/lbuild#13. Hmmm...

**Comment by rleh** [comment]
I can reproduce the same error on my local machine with Windows 10. Fresh installed Win10 (English), Miniconda3 etc. as described in https://modm.io/guide/installation/#windows.
Without changing anything `print(sys.stdout.encoding)` returns `utf-8`.

**Comment by salkinium** [comment]
`lbuild -vvv build` gives you a traceback so you can at least see where it breaks.

**Comment by rleh** [comment]
[Pathlibs `read_text()`](https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_text) causes the issue.
https://github.com/modm-io/modm/search?l=Python&q=read_text%28

**Comment by rleh** [comment]
What the f*ck is wrong with Windows that it takes >5 minutes to download and unzip a 170MB ZIP file?

**Comment by rleh** [comment]
But `lbuild`, `scons` and `arm-none-eabi-gcc` are now somehow running in the Windows machine: https://github.com/rleh/modm/runs/1075398248

But ... what?
```
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
arm-none-eabi-g++ -o D:\a\modm\modm\build\nucleo_f411re\blink\release\main.o -c -fno-exceptions -fno-rtti -fno-unwind-tables -fstrict-enums -fuse-cxa-atexit -std=c++17 -Woverloaded-virtual -fdata-sections -ffunction-sections -finline-limit=10000 -fshort-wchar -fsingle-precision-constant -funsigned-bitfields -funsigned-char -fwrapv -g3 -gdwarf -W -Wall -Wdouble-promotion -Wduplicated-cond -Werror=format -Werror=maybe-uninitialized -Werror=overflow -Werror=sign-compare -Wextra -Wlogical-op -Wpointer-arith -Wundef -Os -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -mthumb -Imodm\ext -Imodm\ext\cmsis\core -Imodm\ext\cmsis\device -Imodm\src -I. main.cpp
main.cpp: In function 'int main()':
main.cpp:34:9: error: 'delay' is not a member of 'modm'
   34 |   modm::delay(Button::read() ? 100ms : 500ms);
      |         ^~~~~
scons: *** [D:\a\modm\modm\build\nucleo_f411re\blink\release\main.o] Error 1
```

**Comment by salkinium** [comment]
> But ... what?

Yeeessssss, I remember this, more path issues: => #310

**Comment by salkinium** [comment]
> What the f*ck is wrong with Windows that it takes >5 minutes to download and unzip a 170MB ZIP file?

macOS CI was also such a PITA taking 8mins or so to setup, nothing beats having a cached Docker container.

**Comment by rleh** [comment]
Maybe using [this "Action"](https://github.com/fiam/arm-none-eabi-gcc) would be faster, but it is currently broken: fiam/arm-none-eabi-gcc#5

**Comment by rleh** [comment]
Linux is so much faster...
grafik

**Comment by salkinium** [comment]
I'm going to add some commits to try and fix the path issues on Windows, but I won't rebase or force push your branch, so we can collaborate on your branch.
Any idea why the GitHub Actions isn't running on modm-io? Do I need to enable more settings?

**Comment by salkinium** [comment]
Also the macOS GitHub Action is faster (~12mins vs ~19mins) than the Travis CI version, so I'm inclined to replace that one.

**Comment by salkinium** [comment]
> macOS GitHub Action is faster than the Travis CI version

This is only because the GitHub Action macOS Image is more up-to-date and therefore Homebrew doesn't need to take as long to update the image.

**Comment by salkinium** [comment]
> What the f*ck is wrong with Windows that it takes >5 minutes to download and unzip a 170MB ZIP file?

I added [`$ProgressPreference = 'SilentlyContinue'`](https://stackoverflow.com/questions/28682642/powershell-why-is-using-invoke-webrequest-much-slower-than-a-browser-download) and now it takes ~7 (!) *seconds* (!!). BWARHARHARHAR, WHY ARE PEOPLE USING WINDOWS?!?

**Comment by rleh** [comment]
> > What the f*ck is wrong with Windows that it takes >5 minutes to download and unzip a 170MB ZIP file?
> 
> I added $ProgressPreference = 'SilentlyContinue' and now it takes ~7 (!) seconds (!!). BWARHARHARHAR, WHY ARE PEOPLE USING WINDOWS?!?

Hahahaha nice!
Unzipping still takes 3 minutes. What is the magic command to speed up unzipping?

**Comment by rleh** [comment]
> I'm going to add some commits to try and fix the path issues on Windows, but I won't rebase or force push your branch, so we can collaborate on your branch.

:+1: 

> Any idea why the GitHub Actions isn't running on modm-io? Do I need to enable more settings?

The workflow file to get merged into this repo. For now we can just use https://github.com/rleh/modm/actions.
As soon as this PR is ready we should change the line `on: [push, pull_request]` to `on: [pull_request]`.

**Comment by salkinium** [comment]
> Unzipping still takes 3 minutes. What is the magic command to speed up unzipping?

Apparently applying this: https://github.com/PowerShell/Microsoft.PowerShell.Archive/issues/32#issuecomment-668732081.

**Comment by rleh** [comment]
> macOS GitHub Action is faster than the Travis CI version
>
> This is only because the GitHub Action macOS Image is more up-to-date and therefore Homebrew doesn't need to take as long to update the image.

In my opinion we can replace Travis CI with GH Actions. Does @strongly-typed have any objections?

**Comment by salkinium** [comment]
Do you @rleh want to continue working on this in your VM? I'm stuck on why `examples_compile.py` returns exit code 1.

**Comment by salkinium** [comment]
I'm also not sure what is needed to be installed for hosted examples and unittests.

**Comment by rleh** [comment]
I will have a look at it, but unfortunately I don't have much time before next week.

> I'm also not sure what is needed to be installed for hosted examples and unittests.

Me neither. Maybe @WasabiFan, @dhebbeker or someone else who uses modm on Windows can help?
If not, I would stay with (only) Cortex-M testing on Windows for now. We can add more test jobs (AVR, hosted, documentation, ...) later.

**Comment by WasabiFan** [comment]
Looking at our Windows setup instructions, I _think_ the only extra dependency for the hosted environment is a native `gcc`; we install it via MinGW.

That being said, we've recently begun a migration to Linux dev environments, because our unit tests are taking dependencies on libs which are easier to install on Linux. The main build still works fine on Windows but we don't maintain test support anymore.

Separately, I'm in awe at the download speed you screenshotted above; our CI container builds (Linux) spend 5-10mins on the toolchain download, and have done so consistently since we initially created those images a year ago. I'll have to look into why yours are so much faster -- perhaps the newer version is hosted on better infrastructure, or perhaps it is a result of the output spew you've disabled.

File: .github/workflows/windows.yml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+jobs:
+  windows_testing:
+    runs-on: windows-2019
+    env:
+      PYTHONIOENCODING: "utf-8"
+
+    steps:
+
+      # Disabling snake-oil for performance reasons
```
Comment (dhebbeker) [inline .github/workflows/windows.yml]: :rofl:

**Comment by rleh** [comment]
MinGW: :heavy_check_mark: 

But the  hosted tests fail on windows:
```
lbuild -c config/hosted.xml -D":target=hosted-windows" -p ../build/generated-unittest/hosted build --no-log
[ERROR] user.modm.platform.core: Assertions are not implemented!
[ERROR] user.modm.platform.core: Delay functions are not implemented!
>> modm: Recomputing device cache...

scons -C ../build/generated-unittest/hosted run
scons: Entering directory `D:\a\modm\modm\build\generated-unittest\hosted'
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
Create Tests··· unittest_runner.cpp
Compiling C++·· release\unittest_runner.o
In file included from modm\src/modm/processing/timer/timeout.hpp:21,
                 from modm\src/modm/processing/timer.hpp:15,
                 from modm\src/modm/communication/xpcc/dispatcher.hpp:21,
                 from D:\a\modm\modm\build\generated-unittest\hosted\modm-test\src\modm-test\communication\xpcc\dispatcher_test.hpp:20,
                 from unittest_runner.cpp:25:
modm\src/modm/architecture/interface/assert.hpp:172:10: fatal error: modm/platform/core/assert_impl.hpp: No such file or directory
  172 | #include <modm/platform/core/assert_impl.hpp>
```

**Comment by salkinium** [comment]
> fatal error: modm/platform/core/assert_impl.hpp: No such file or directory

Yes, the implementation is missing for this, since I couldn't find out how to access the custom linker section that contains the assertion handlers on Windows. So this is not a quick bug fix, we'll probably have to leave it failing for now.

**Comment by salkinium** [comment]
> So this is not a quick bug fix, we'll probably have to leave it failing for now.

Actually, let me patch this with a dummy implementation. Assertions aren't that important on Hosted anyways and especially not for the unittests.

Do you want to add the `modm/example/windows` compilation back in? 😛

**Comment by rleh** [comment]
> Do you want to add the modm/example/windows compilation back in?

Yes. But how...? f1bb2a7 seems not to work.

**Comment by salkinium** [comment]
> Yes. But how...? f1bb2a7 seems not to work.

No, generic won't work for windows, it's too different. You can only add the `examples/windows` folder.

**Comment by salkinium** [comment]
Well… assertions have been "fixed", onto the next issue…
```
modm\src\info_build.c:16:35: warning: missing terminating " character
 const char *MODM_BUILD_COMPILER = "g++ (MinGW.org GCC Build-2) 9.2.0 ";
```

**Comment by rleh** [comment]
> Well… assertions have been "fixed"

[src/modm/platform/core/hosted/module.lb](https://github.com/modm-io/modm/blob/develop/src/modm/platform/core/hosted/module.lb#L48-L52)

**Comment by salkinium** [comment]
Fixed an a trailing `\r` in the info tool, the Regex parsing the compiler version seems to be wrong.

`example_compile.py` seems to not properly wait for the multiprocessing pool to end, thus it's calling all lbuild commands, but then only compiling *some* of them, thus failing, since not all examples were compiled (duh)… wtf.

Now there's something, something with <inttypes.h>.

```
modm\src\modm\architecture\interface\can_message.cpp:40:21: error: expected ')' before 'PRIx32'
  s.printf("id = %04" PRIx32 ", len = ", m.identifier);
          ~          ^~~~~~~
                     )
modm\src\modm\architecture\interface\can_message.cpp:40:19: error: conversion lacks type at end of format [-Werror=format=]
  s.printf("id = %04" PRIx32 ", len = ", m.identifier);
                   ^
modm\src\modm\architecture\interface\can_message.cpp:40:11: error: too many arguments for format [-Werror=format-extra-args]
  s.printf("id = %04" PRIx32 ", len = ", m.identifier);
           ^~~~~~~~~~
```

**Comment by rleh** [comment]
https://stackoverflow.com/questions/14535556/why-doesnt-priu64-work-in-this-code/18719205#18719205

**Comment by salkinium** [comment]
Already fixed the inttypes, also have some idea about the multiprocessing issue: https://stackoverflow.com/questions/20360686/compulsory-usage-of-if-name-main-in-windows-while-using-multiprocessi

**Comment by salkinium** [comment]
Windows doesn't have `os.fork()` support, so multiprocessing is implemented using weird hacks… I've disabled it for now, so we'll have to generate the examples single-threaded, which is okish considering the alternatives.

**Comment by salkinium** [comment]
For god's sake, WINDOWS!
```
In file included from modm\src\modm\io\iostream_printf.cpp:27:
modm\ext/printf/printf.source:580:12: error: conflicting declaration of C function 'int _vsnprintf(out_fct_type, char*, size_t, const char*, va_list)'
 static int _vsnprintf(out_fct_type out, char* buffer, const size_t maxlen, const char* format, va_list va)
            ^~~~~~~~~~
In file included from modm\src\modm\io\iostream_printf.cpp:12:
c:\mingw-get\include\stdio.h:526:38: note: previous declaration 'int _vsnprintf(char*, size_t, const char*, char*)'
 _CRTIMP __cdecl __MINGW_NOTHROW  int _vsnprintf (char *, size_t, const char *, __VALIST);
                                      ^~~~~~~~~~
```

**Comment by salkinium** [comment]
Another difference on Windows:

```
Assembling····· build\release\modm\ext\crashcatcher\src\CrashCatcher_armv7m.o

scons: building terminated because of errors.
arm-none-eabi-as: option `-gdwarf' is ambiguous
```

**Comment by salkinium** [comment]
I've never felt more professional than when setting up a CI by basically just implementing suggestions of StackOverflow. This really is the future of programming…

**Comment by salkinium** [comment]
Honestly a bit shocked how broken our Windows support is, and how few Windows users complained. We either don't have a lot of Windows users, or they are savy enough to fix the issues themselves, but then not contribute back…

**Comment by salkinium** [comment]
Why is the assembler so wonky on Windows?!?
```
Assembling····· build\release\modm\ext\crashcatcher\src\CrashCatcher_armv7m.o
scons: building terminated because of errors.
modm\ext\crashcatcher\src\CrashCatcher_armv7m.S: Assembler messages:
modm\ext\crashcatcher\src\CrashCatcher_armv7m.S:47: Error: syntax error -- `ldr sp,=(g_crashCatcherStack+4*CRASH_CATCHER_STACK_WORD_COUNT)'
modm\ext\crashcatcher\src\CrashCatcher_armv7m.S:79: Error: selected FPU does not support instruction -- `vmrs r1,fpscr'
modm\ext\crashcatcher\src\CrashCatcher_armv7m.S:82: Error: selected FPU does not support instruction -- `vstmia.32 r0!,{s0-s31}'
modm\ext\crashcatcher\src\CrashCatcher_armv7m.S:106: Error: selected FPU does not support instruction -- `vstmia.32 r0!,{s16-s31}'
```

Update: So the `.S` files are supposed to be processed by the preprocessor, then assembled, hence the `Error: syntax error -- `ldr sp,=(g_crashCatcherStack+4*CRASH_CATCHER_STACK_WORD_COUNT)'` issue.

**Comment by salkinium** [comment]
Oh interesting, the Windows CI calls `arm-none-eabi-as`:

```
arm-none-eabi-as -g3 -gdwarf-3 -mcpu=cortex-m3 -mthumb -o D:\a\modm\modm\build\nucleo_f103rb\hard_fault\release\modm\ext\crashcatcher\src\CrashCatcher_armv7m.o modm\ext\crashcatcher\src\CrashCatcher_armv7m.S
```

But my local version calls `arm-none-eabi-gcc`:

```
arm-none-eabi-gcc -g3 -gdwarf-3 -mcpu=cortex-m3 -mthumb -Imodm/ext -Imodm/ext/cmsis/core -Imodm/ext/cmsis/device -Imodm/ext/crashcatcher/include -Imodm/src -Imodm -c -o modm/build/nucleo_f103rb/hard_fault/release/modm/ext/crashcatcher/src/CrashCatcher_armv7m.o modm/ext/crashcatcher/src/CrashCatcher_armv7m.S
```

**Comment by salkinium** [comment]
Apparently `.S` files are only preprocessed on POSIX, only `.sx` files are preprocessed on all platforms…

**Comment by salkinium** [comment]
\o/
Cortex-M tests are passing!
Hosted unittest is compiling and almost passing (except for 3 tests)!

**Comment by salkinium** [comment]
Something to do with inttypes or I enabled the wrong overload in `IOStream.hpp`:
```
>>> io_stream_test
FAIL: io_stream_test:621 : 
[F, E, D, C, B, A, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
[I, 6, 4, X, , , , , , , , , , , , ]
FAIL: io_stream_test:626 : 
[-, 9, 2, 2, 3, 3, 7, 2, 0, 3, 6, 8, 5, 4, 7, 7, 5, 8, 0, 6]
[I, 6, 4, d, , , , , , , , , , , , , , , , ]
```

Not sure, probably related to the float comparison sigma.
```
>>> vector2_test
FAIL: vector2_test:300 : 1.28171e+02 == 1.28171e+02
```

**Comment by WasabiFan** [comment]
> Honestly a bit shocked how broken our Windows support is, and how few Windows users complained. We either don't have a lot of Windows users, or they are savy enough to fix the issues themselves, but then not contribute back…

This is why we [run on an old patched build](https://github.com/uw-advanced-robotics/modm) 😉 I think you've already figured out most of the applicable fixes anyway (e.g. the `vsprintf` BS). There were lots of portability quirks that had to be discovered through trial-and-error, but once those were patched over for Widows-specific issues, modm as a platform worked great, even when transiting between OSes.

**Comment by salkinium** [comment]
> Something to do with inttypes or I enabled the wrong overload in IOStream.hpp:

Nope, something much more stupid: https://stackoverflow.com/a/55099671

**Comment by salkinium** [comment]
I added the ability to run the hosted examples via `examples_compile.py`. So now the examples/linux get compiled and run in the CI on Linux, macOS and Windows (except the serial examples)! \o/

**Comment by salkinium** [comment]
All Github CI's are passing, is it ok with you @rleh, if I rebase and regroup the commits for a review?

**Comment by salkinium** [comment]
>  I rebase and regroup the commits for a review?

I assumed yes.

I'm happy with this setup, let's review this.

File: .github/workflows/macos.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+name: Run tests on MacOS
+
+on: [push, pull_request]
```
Comment (salkinium) [inline .github/workflows/macos.yml]:
```suggestion
on: [pull_request]
```

File: .github/workflows/windows_armcortexm.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
+name: Run ARM Cortex-M tests on Windows
+
+on: [push, pull_request]
```
Comment (salkinium) [inline .github/workflows/windows_armcortexm.yml]:
```suggestion
on: [pull_request]
```

File: .github/workflows/windows_hosted.yml [hunk 4]
```diff
@@ -0,0 +1,68 @@
+name: Run hosted tests on Windows
+
+on: [push, pull_request]
```
Comment (salkinium) [inline .github/workflows/windows_hosted.yml]:
```suggestion
on: [pull_request]
```

File: examples/linux/assert/project.xml [hunk 5]
```diff
@@ -1,4 +1,5 @@
 <library>
+  <!-- CI: run fail -->
```
Comment (salkinium) [inline examples/linux/assert/project.xml:2]: This is a hint to the `examples_compile.py` script that the example should be run and that a failure is expected.

File: ext/adamgreen/catcher.lb [hunk 6]
```diff
@@ -34,5 +34,7 @@ def build(env):
     env.outbasepath = "modm/ext/crashcatcher"
 
     env.copy("crashcatcher/CrashCatcher/include/CrashCatcher.h", "include/CrashCatcher.h")
-    ignore = "*armv7m.S" if "m0" in core else "*armv6m.S"
-    env.copy("crashcatcher/CrashCatcher/Core/src", "src", ignore=env.ignore_files(ignore))
+    env.copy("crashcatcher/CrashCatcher/Core/src", "src", ignore=env.ignore_files("*.S"))
+    version = "armv{}m".format("6" if "m0" in core else "7")
+    env.copy("crashcatcher/CrashCatcher/Core/src/CrashCatcher_{}.S".format(version),
+             "src/CrashCatcher_{}.sx".format(version))
```
Comment (salkinium) [inline ext/adamgreen/catcher.lb:40]: I should open an issue for this in the original repo.

File: repo.lb [hunk 7]
```diff
@@ -228,10 +228,15 @@ def init(repo):
             path = normpath(path)
             escaped = "\\" * (2 ** escape_level)
             return path.replace("\\", escaped)
+        def posixify(path):
+            return normpath(path).replace("\\", "/")
```
Comment (salkinium) [inline repo.lb:232]: I feel like this is a dumb solution, and that lbuild should have some option to reformat paths somehow, but it's all such a horrible mess, I don't want to deal with it.

File: src/modm/debug/logger/level.hpp [hunk 8]
```diff
@@ -15,8 +15,7 @@
 #ifndef MODM_LOG_LEVEL_HPP
 #define MODM_LOG_LEVEL_HPP
 
-#pragma push_macro("ERROR") // avoid collision with ERROR defined macro in winsock.h
-#undef ERROR
+#undef ERROR // avoid collision with ERROR defined macro in winsock.h
```
Comment (salkinium) [inline src/modm/debug/logger/level.hpp:18]: push/pop wasn't enough, not sure how to solve this properly.

File: src/modm/platform/core/cortex/assert.cpp.in [hunk 9]
```diff
@@ -31,6 +31,10 @@ extern AssertionHandler __assertion_table_end __asm("section$end$__DATA$modm_ass
 %% elif target.family == "linux"
 extern AssertionHandler __assertion_table_start __asm("__start_modm_assertion");
 extern AssertionHandler __assertion_table_end __asm("__stop_modm_assertion");
+%% elif target.family == "windows"
+// FIXME: Figure out how to access custom linker sections
+static AssertionHandler __assertion_table_start;
+#define __assertion_table_end __assertion_table_start
```
Comment (salkinium) [inline src/modm/platform/core/cortex/assert.cpp.in:37]: Don't know how to define and access custom linker sections on Windows, I guess it's not that important anyways.

File: test/modm/io/io_stream_test.cpp [hunk 10]
```diff
@@ -624,6 +625,7 @@ IoStreamTest::testPrintf3()
 	(*stream).printf("%lld", longlong);
 	TEST_ASSERT_EQUALS_ARRAY("-9223372036854775806", device.buffer, 20);
 	(*stream).flush();
+#endif
```
Comment (salkinium) [inline test/modm/io/io_stream_test.cpp:628]: This was supposed to work by defining `__USE_MINGW_ANSI_STDIO`, but you're not supposed to define it directly, instead use something like `_POSIX_C_SOURCE`, but neither ones worked. Meh.

File: test/modm/math/geometry/vector2_test.cpp [hunk 11]
```diff
@@ -296,8 +296,10 @@ Vector2Test::testRotateFloat()
 
 	a.rotate(modm::Angle::toRadian(20));
 
-	TEST_ASSERT_EQUALS_FLOAT(a.getX(), 59.76724775f);
-	TEST_ASSERT_EQUALS_FLOAT(a.getY(), 128.1712764f);
+	TEST_ASSERT_EQUALS_FLOAT(a.getX(), 59.767247746f);
+#ifndef MODM_OS_WIN32 // FIXME: Windows has some unknown accuracy issue here
+	TEST_ASSERT_EQUALS_FLOAT(a.getY(), 128.1712764112f);
+#endif
```
Comment (salkinium) [inline test/modm/math/geometry/vector2_test.cpp:302]: No clue what the issue is, perhaps it's casting to double here or something.

**Review by salkinium — APPROVED** [review]

File: tools/build_script_generator/common.py [hunk 12]
```diff
@@ -217,10 +217,10 @@ def common_compiler_flags(compiler, target):
         "-funsigned-char",
         "-fwrapv",
         # "-fmerge-all-constants",
-
         "-g3",
-        "-gdwarf",
+        "-gdwarf-3"
```
Comment (rleh) [inline tools/build_script_generator/common.py]:
```suggestion
        "-gdwarf-3",
```

File: tools/build_script_generator/common.py [hunk 13]
```diff
@@ -264,7 +264,7 @@ def common_compiler_flags(compiler, target):
     # flags only for Assembly
     flags["asflags"] = [
         "-g3",
-        "-gdwarf",
+        "-gdwarf-3"
```
Comment (rleh) [inline tools/build_script_generator/common.py]:
```suggestion
        "-gdwarf-3",
```

**Comment by rleh** [comment]
I'm trying to add a runner with AArch64/arm64 architecture and linux, currently only available on Travis.

Seems to catch two more issues.

Hosted:
```
modm/src/modm/architecture/detect.hpp:228:3: error: #error "Couldn't determine target CPU!"
228 | # error "Couldn't determine target CPU!"
```

ARM Cortex-M:
```
Linking········ build/release/blink.elf
scons: building terminated because of errors.`_sbrk_r' referenced in section `.text' of /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc_nano.a(lib_a-nano-mallocr.o): defined in discarded section `.Heap_is_not_implemented!__Please_include_the__modm:platform:heap__module_in_your_project!' of build/release/modm/libmodm.a(no_heap.o)
`_sbrk_r' referenced in section `.text' of /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc_nano.a(lib_a-nano-mallocr.o): defined in discarded section `.Heap_is_not_implemented!__Please_include_the__modm:platform:heap__module_in_your_project!' of build/release/modm/libmodm.a(no_heap.o)
`_sbrk_r' referenced in section `.text' of /usr/lib/gcc/arm-none-eabi/9.2.1/../../../arm-none-eabi/lib/thumb/v7-m/nofp/libc_nano.a(lib_a-nano-mallocr.o): defined in discarded section `.Heap_is_not_implemented!__Please_include_the__modm:platform:heap__module_in_your_project!' of build/release/modm/libmodm.a(no_heap.o)
collect2: error: ld returned 1 exit status
```

@salkinium Any ideas?

Maybe I should move the Aarch64-Runner to another Pull-Request...

**Comment by rleh** [comment]
> Remove PYTHONIOENCODING, because its not necessary? (5268d8d)

`PYTHONIOENCODING` does not seem to be necessary, tests have been successful.

**Comment by salkinium** [comment]
`modm/architecture/detect.hpp` needs to be adapted to set the right macro defines.

Regarding Cortex-M: Perhaps there is an issue with weak linking or sections garbage collection (`-Wl,--gc-sections`)on the Arm64 toolchain.

> Maybe I should move the Aarch64-Runner to another Pull-Request...

Yes, I would suggest so.

**Review by rleh — COMMENTED** [review]
Approve :heavy_check_mark: 
(I'm not allowed to approve _my own_ pull request :crying_cat_face:)

**Comment by rleh** [comment]
>> Maybe I should move the Aarch64-Runner to another Pull-Request...
>
> Yes, I would suggest so.

Done: #467

**Comment by rleh** [comment]
I squashed my fixup-commits (again), made both of us authors of the commits and rebased onto develop.
@salkinium Ready to merge!

**Pull request merged**